dropshot/tests/test_versions.rs

```diff
@@ -0,0 +1,521 @@
+        TestCase::new(
+            Method::PUT,
+            "/demo",
+            Some("0.9.0"),
+            StatusCode::OK,
+            HANDLER4_MSG,
+        ),
+        TestCase::new(
+            Method::PUT,
+            "/demo",
+            Some("1.0.0"),
+            StatusCode::OK,
+            HANDLER4_MSG,
+        ),
+        TestCase::new(
+            Method::PUT,
+            "/demo",
+            Some("1.1.0"),
+            StatusCode::OK,
+            HANDLER4_MSG,
+        ),
+        TestCase::new(
+            Method::PUT,
+            "/demo",
+            Some("1.3.1"),
+            StatusCode::OK,
+            HANDLER4_MSG,
+        ),
+        TestCase::new(
+            Method::PUT,
+            "/demo",
+            Some("1.4.0"),
+            StatusCode::OK,
+            HANDLER4_MSG,
+        ),
+        TestCase::new(
+            Method::PUT,
+            "/demo",
+            Some("1.5.0"),
+            StatusCode::BAD_REQUEST,
+            "server does not support this API version: 1.5.0",
+        ),
+    ];
+
+    for t in test_cases {
+        println!("test case: {:?}", t);
+        let mut request = client.request(t.method, mkurl(t.path));
+        if let Some(h) = t.header {
+            request = request.header(VERSION_HEADER_NAME, h);
+        }
+        let response = request.send().await.unwrap();
+        assert_eq!(response.status(), t.expected_status);
+
+        if !t.expected_status.is_success() {
+            let error: HttpErrorResponseBody = response.json().await.unwrap();
+            assert_eq!(error.message, t.body_contents);
+        } else {
+            // This is ugly.  But it's concise!
+            //
+            // We want to use a different type for `handler1` just so that we
+            // can check in test_versions_openapi() that it appears in the
+            // OpenAPI spec only in the appropriate versions.  So if we're
+            // expecting to get something back from handler1, then parse the
+            // type a little differently.
+            let body: String = if t.body_contents == HANDLER1_MSG {
+                let body: EarlyReturn = response.json().await.unwrap();
+                body.real_message
+            } else {
+                response.json().await.unwrap()
+            };
+
+            assert_eq!(body, t.body_contents);
+        }
+    }
+}
+
+/// Test that the generated OpenAPI spec only refers to handlers in that version
+/// and types that are used by those handlers.
+#[test]
+fn test_versions_openapi() {
+    let api = api();
+
+    for version in ["0.9.0", "1.0.0", "1.1.0", "1.3.1", "1.4.0"] {
+        let semver: semver::Version = version.parse().unwrap();
+        let mut found = Cursor::new(Vec::new());
+        api.openapi("Evolving API", semver).write(&mut found).unwrap();
+        let actual = std::str::from_utf8(found.get_ref()).unwrap();
+        expectorate::assert_contents(
+            &format!("tests/test_openapi_v{}.json", version),
```

Comment:
what do you think about 
```suggestion
            &format!("tests/test_versions_v{}.json", version),
```
or something to help associate json with rs?

Reply:
Done.